New Feature: Lookup Boxes by app-id + Name and Search Boxes by app-id (#1117)
algorand/indexer: api/README.md

## Why do we have indexer.oas2.yml and indexer.oas3.yml?

We chose to maintain V2 and V3 versions of the spec because OpenAPI v3 doesn't seem to be widely supported. Some tools worked better with V3 and others with V2, so having both available has been useful. To reduce developer burdon, the v2 specfile is automatically converted v3 using [converter.swagger.io](http://converter.swagger.io/).

> # Fixtures Test

Reviewer: Thanks for adding this!

## What is a **Fixtures Test**?

Currently (September 2022) [fixtures_test.go](./fixtures_test.go) is a library that allows testing Indexer's router to verify that endpoints accept parameters and respond as expected, and guard against future regressions. [app_boxes_fixtures_test.go](./app_boxes_fixtures_test.go) is an example _fixtures test_ and is the _creator_ of the fixture [boxes.json](./test_resources/boxes.json).

A fixtures test

1. is defined by a go-slice called a _Seed Fixture_  e.g. [var boxSeedFixture](https://github.com/algorand/indexer/blob/b5025ad640fabac0d778b4cac60d558a698ed560/api/app_boxes_fixtures_test.go#L302-L692) which contains request information for making HTTP requests against an Indexer server
2. iterates through the slice, making each of the defined requests and generating a _Live Fixture_
3. reads a _Saved Fixture_ from a json file e.g. [boxes.json](./test_resources/boxes.json)
4. persists the _Live Fixture_ to a json file not in source control
5. asserts that the _Saved Fixture_ is equal to the _Live Fixture_

> In reality, because we always want to save the _Live Fixture_ before making asserstions that could fail the test and preempt saving, steps (3) and (4) happen in the opposite order.

Author: ```suggestion
In reality, because we always want to save the _Live Fixture_ before making assertions that could fail the test and pre-empt saving, steps (3) and (4) happen in the opposite order.
```

## What's the purpose of a Fixtures Test?

A fixtures test should allow one to quickly stand up an end-to-end test to validate that Indexer endpoints are working as expected. After Indexer's state is programmatically set up, it's easy to add new requests and verify that the responses look exactly as expected. Once you're satisfied that the responses are correct, it's easy to _freeze_ the test and guard against future regressions.
## What does a **Fixtures Test Function** Look Like?

[func TestBoxes](https://github.com/algorand/indexer/blob/b5025ad640fabac0d778b4cac60d558a698ed560/api/app_boxes_fixtures_test.go#L694_L704) shows the basic structure of a fixtures test.

> 1. `setupLiveServerAndReturnShutdownFunc()` is called to set up the Indexer database

Author: ```suggestion
1. `setupIdbAndReturnShutdownFunc()` is called to set up the Indexer database
```

  * this isn't expected to require modification

>   2. `setupLiveBoxes()` is used to prepare the local ledger and process blocks in order to bring Indexer into a particular state

Reviewer: ```suggestion
2. `setupLiveBoxes()` is used to prepare the local ledger and process blocks in order to bring Indexer into a particular state
```

  * this will always depend on what the test is trying to achieve
  * in this case, an app was used to create and modify a set of boxes which are then queried against
  * it is conceivable that instead of bringing Indexer into a particular state, the responses from the DB or even the handler may be mocked, so we could have had `setupLiveBoxesMocker()` instead of `setupLiveBoxes()`
3. `setupLiveServerAndReturnShutdownFunc()` is used to bring up an instance of a real Indexer.
  * this shouldn't need to be modified; however, if running in parallel and making assertions that conflict with other tests, you may need to localize the variable `fixtestListenAddr` and run on a separate port
  * if running a mock server instead, a different setup function would be needed
4. `validateLiveVsSaved()` runs steps (1) through (5) defined in the previous section
  * this is designed to be generic and ought not require much modification going forward

## Which Endpoints are Currently _Testable_ in a Fixtures Test?

> Endpoints defined in [proverRoutes](https://github.com/algorand/indexer/blob/b955a31b10d8dce7177383895ed8e57206d69f67/api/fixtures_test.go#L213) are testable.

Author: ```suggestion
Endpoints defined in [proverRoutes](https://github.com/algorand/indexer/blob/b955a31b10d8dce7177383895ed8e57206d69f67/api/fixtures_test.go#L232-L263) are testable.
```
